File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const res = [];

    function backtrack(currNums, accum, index) {
      if (accum < 0) return;
      else if (accum === 0) {
        res.push(currNums);
        return;
      };

      // need to account for index here so duplicates won't be processed, starts from the next one every time.
      for (let i = index; i < candidates.length; i++) {
        backtrack([...currNums, candidates[i]], accum - candidates[i], index++)
      };
    };
    if (candidates.length) backtrack([], target, 0);

    return res;
};  

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        
        def backtrack(remainder, path):
            if remainder == 0:
                p = path[:]
                p.sort()
                if p not in res:
                    res.append(p)
                return

            if remainder < 0:
                return

            for num in candidates:

                new_path = path[:]
                new_path.append(num)

                backtrack(remainder - num, new_path)
            
        backtrack(target, [])
        return res
                

